lib/src/main/java/org/kpipe/consumer/OffsetManager.java

```diff
@@ -0,0 +1,574 @@
+  }
+
+  /** Performs the actual commit operation and cleans up if successful. */
+  private boolean performCommitAndCleanup(Map<TopicPartition, OffsetAndMetadata> offsetsToCommit, int timeoutSeconds)
+    throws InterruptedException {
+    final var future = new CompletableFuture<Boolean>();
+    enqueueCommand(
+      () -> {
+        try {
+          consumer.commitSync(offsetsToCommit);
+          future.complete(true);
+        } catch (final Exception e) {
+          LOGGER.log(WARNING, "Failed to commit offsets", e);
+          future.complete(false);
+        }
+      },
+      "commitSync"
+    );
+
+    try {
+      final var result = future.get(timeoutSeconds, TimeUnit.SECONDS);
+      if (result) {
+        cleanupCommittedOffsets(offsetsToCommit);
+      }
+      return result;
+    } catch (final ExecutionException | TimeoutException e) {
+      LOGGER.log(WARNING, "Error waiting for offset commit", e);
+      return false;
+    }
+  }
+
+  /**
+   * Cleans up offsets that have been successfully committed to prevent memory leaks.
+   *
+   * @param committedOffsets Map of offsets that were successfully committed
+   */
+  private void cleanupCommittedOffsets(final Map<TopicPartition, OffsetAndMetadata> committedOffsets) {
+    committedOffsets.forEach((partition, metadata) -> {
+      long committedOffset = metadata.offset();
+      final var processed = processedOffsets.get(partition);
+      if (processed != null) {
+        try {
+          // Get subSet view of elements less than committedOffset and clear it
+          final var toRemove = processed.headSet(committedOffset);
+          toRemove.clear();
+        } catch (final UnsupportedOperationException e) {
+          // Fall back to standard removeIf if headSet clear isn't supported
+          processed.removeIf(offset -> offset < committedOffset);
+        }
+
+        if (processed.isEmpty()) {
+          processedOffsets.remove(partition);
+        }
+      }
+    });
+  }
+
+  /**
+   * Updates the committable offset for a partition based on processed offsets. Handles both
+   * contiguous and non-contiguous offset tracking.
+   *
+   * @param partition The partition to update
+   */
+  public void updateCommittableOffset(final TopicPartition partition) {
+    final var currentPosition = nextOffsetsToCommit.get(partition);
+    if (currentPosition == null) {
+      return;
+    }
+
+    final var pending = pendingOffsets.get(partition);
+    final var processed = processedOffsets.get(partition);
+    if (processed == null || processed.isEmpty()) {
+      return;
+    }
+
+    // Find the highest contiguous processed offset
+    final var highestContiguous = findHighestContiguousOffset(processed);
+    final var nextExpected = highestContiguous + 1;
+
+    // If all offsets are processed, commit one past the highest contiguous
+    if (pending == null || pending.isEmpty()) {
+      nextOffsetsToCommit.put(partition, nextExpected);
+      return;
+    }
```

Comment:
this seems wrong? the processed offsets are already holding offset +1 as values - next expected is +1 again.
So lets say we processed offsets 0-9 inclusive - what actually is stored in processedOffsets is 1-10 inclusive and then nextExpected=10+1 - so 11... but it should commit offset 10 (last processed offset of 9 +1 ).

Reply:
I am not sure why having pending offsets has an impact here? 
Lets say offsets polled and submitted are 0-9 inclusive
processed 0-5, 
pending 6,7,
processed 8,9
logic should just commit 6 (as highest consecutive processed of 5 +1 ) regardless of pending offsets?

Reply:
removed

Reply:
See this [comment](https://github.com/eschizoid/kpipe/pull/51#discussion_r2070923072)